model-optimizer/extensions/back/InterpolateReshape.py

```diff
@@ -0,0 +1,150 @@
+        if len(non_interp_concat_srcs) == 0:
+            return
+
+        graph = node.graph
+        src = non_interp_concat_srcs[0]
+
+        shape = Shape(graph, {'name': src.node.soft_get('name', src.node.id) + '/Shape'}).create_node()
+        shape.in_port(0).connect(src)
+        gather = create_op_with_const_inputs(graph, Gather, {1: np.array(node.axes, dtype=np.int32), 2: int64_array(0)},
+                                             {'name': shape.name + '/Gathered'}, shape)
+        node.in_port(1).get_connection().set_source(gather.out_port(0))
+
+    def find_and_replace_pattern(self, graph: Graph):
+        for node in graph.get_op_nodes(type='Interpolate'):
+            if node.in_port(1).get_source().node.soft_get('type') != 'Const':
+                continue
+            dsts = node.out_port(0).get_destinations()
+            if len(dsts) == 1 and dsts[0].node.soft_get('type') == 'Concat':
+                self.make_interpolate_reshapeable(node)
+
+
+class InterpolateReshapeWA(BackReplacementPattern):
+    """
+    Replaces hard-coded 1-port input of Interpolate with reshape-able sub-graph.
+    WARNING: Could cause troubles if model has hard-coded Interpolate intentionally -- rare situation
+
+    BEFORE:
+        input                   Const
+    shape=[1, 3, 30, 40]      value=[60, 160]
+            \                   /
+           Interpolate(axes=(2, 3))
+            shape=[1, 3, 60, 160]
+
+    AFTER:
+            input
+    shape=[1, 3, 30, 40]
+        |                \
+        |              ShapeOf
+        |                |
+        |              Gather                Const
+        |        indices=(2, 3); axis=0    value=[2, 4]
+        |                \                /
+        |                    Multiply
+        |                   /
+    Interpolate(axes=(2, 3))
+      shape=[1, 3, 60, 140]
+    """
+    enabled = False
+    graph_condition = [lambda graph: graph.graph['cmd_params'].generate_experimental_IR_V10,
+                       lambda graph: graph.graph['cmd_params'].keep_shape_ops]
+    force_clean_up = True
+
+    def run_after(self):
+        return [InterpolateConcat]
+
+    @staticmethod
+    def make_interpolate_reshapeable(node):
+        axes = node.axes
+        input_shape = node.in_port(0).data.get_shape()
+        output_shape = node.out_port(0).data.get_shape()
+        if not np.all(np.remainder(output_shape, input_shape) == 0):
+            return
+        graph = node.graph
+        name = node.soft_get('name', node.id)
+        shape = Shape(graph, {'name': name + '/ShapeOf'}).create_node()
+        shape.in_port(0).connect(node.in_port(0).get_source())
+        gather = create_op_with_const_inputs(graph, Gather, {1: np.array(axes, dtype=np.int32), 2: int64_array(0)},
+                                             {'name': shape.name + '/Gathered'}, shape)
+        multipliers = int64_array(output_shape[axes] / input_shape[axes])
```

Comment:
can multipliers be not integers? what if output H and W are not divisible by input H and W.

Reply:
please insert a check that output H,W are evenly divisible by input H,W

Reply:
1. we decided to keep it classy and fix only the models that worth saving. otherwise -- who knows what could happen? different rounding and all that messy things. real cases are usually this simple -- just having 2x for input spatial dims.
2. in is there on the line 135